diff --git a/drivers/net/ethernet/mediatek/mtk_eth_soc.c b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
index 29b4aea..03759f7 100644
--- a/drivers/net/ethernet/mediatek/mtk_eth_soc.c
+++ b/drivers/net/ethernet/mediatek/mtk_eth_soc.c
@@ -2104,6 +2104,7 @@ static int mtk_poll_rx(struct napi_struct *napi, int budget,
 
 			if (ret != XDP_PASS)
 				goto skip_rx;
+			net_prefetch(xdp.data_meta);
 
 			skb = build_skb(data, PAGE_SIZE);
 			if (unlikely(!skb)) {

